src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs

```diff
@@ -0,0 +1,189 @@
+            private static readonly ConcurrentDictionary<Type, Type> _metadataTypeCache = new ConcurrentDictionary<Type, Type>();
+
+            // Stores the attributes for a member info
+            private static readonly ConcurrentDictionary<Tuple<Type, string>, Attribute[]> _typeMemberCache = new ConcurrentDictionary<Tuple<Type, string>, Attribute[]>();
+
+            // Stores whether or not a type and associated metadata type has been checked for validity
+            private static readonly ConcurrentDictionary<Tuple<Type, Type>, bool> _validatedMetadataTypeCache = new ConcurrentDictionary<Tuple<Type, Type>, bool>();
+
+            public static void ValidateMetadataType(Type type, Type associatedType)
+            {
+                Tuple<Type, Type> typeTuple = new Tuple<Type, Type>(type, associatedType);
+                if (!_validatedMetadataTypeCache.ContainsKey(typeTuple))
+                {
+                    CheckAssociatedMetadataType(type, associatedType);
+                    _validatedMetadataTypeCache.TryAdd(typeTuple, true);
+                }
+            }
+
+            public static Type GetAssociatedMetadataType(Type type)
+            {
+                Type associatedMetadataType = null;
+                if (_metadataTypeCache.TryGetValue(type, out associatedMetadataType))
+                {
+                    return associatedMetadataType;
+                }
+
+                // Try association attribute
+                MetadataTypeAttribute attribute = (MetadataTypeAttribute)Attribute.GetCustomAttribute(type, typeof(MetadataTypeAttribute));
+                if (attribute != null)
+                {
+                    associatedMetadataType = attribute.MetadataClassType;
+                }
+                _metadataTypeCache.TryAdd(type, associatedMetadataType);
+                return associatedMetadataType;
+            }
+
+            private static void CheckAssociatedMetadataType(Type mainType, Type associatedMetadataType)
+            {
+                // Only properties from main type
+                HashSet<string> mainTypeMemberNames = new HashSet<string>(mainType.GetProperties().Select(p => p.Name));
+
+                // Properties and fields from buddy type
+                var buddyFields = associatedMetadataType.GetFields().Select(f => f.Name);
+                var buddyProperties = associatedMetadataType.GetProperties().Select(p => p.Name);
+                HashSet<string> buddyTypeMembers = new HashSet<string>(buddyFields.Concat(buddyProperties), StringComparer.Ordinal);
+
+                // Buddy members should be a subset of the main type's members
+                if (!buddyTypeMembers.IsSubsetOf(mainTypeMemberNames))
+                {
+                    // Reduce the buddy members to the set not contained in the main members
+                    buddyTypeMembers.ExceptWith(mainTypeMemberNames);
+
+                    throw new InvalidOperationException(String.Format(
```

Comment:
Please use `SR.Format` when formatting an exception string that came from `SR.xxxx`.
The reason relates to AOT compilation, it can strip exception messages, but we still want the parameters. String.Format would fail.